File: Difficulty: Medium/Stock span problem/stock-span-problem.cpp
class Solution {
  public:
    vector<int> calculateSpan(vector<int>& arr) {
        // code here
        stack<pair<int,int>>s;
        int n = arr.size();
        vector<int>ans(n);
        int index = -1;
        for(int i = 0 ; i< n ; i++){
            index+=1;
            
            while(!s.empty() && s.top().first<=arr[i])s.pop();
            
            int res = index - (s.empty()?-1:s.top().second);
            ans[i]=res;
            s.push({arr[i],index});
            
        }
        return ans;
    }
};

File: Difficulty: Medium/Stock Span Problem/stock-span-problem.py
from collections import deque
class Solution:
    def calculateSpan(self, arr):
        # code here
        
        st=deque()
        index=0
        ans=[]
        
        for a in arr:
            
            while st and st[-1][0]<=a:
                st.pop()
            
            if not st:
                ans.append(index+1)
            else:
                ans.append(index-st[-1][1])
            st.append((a,index))
            index+=1
            
        return ans
                
            
            